### PR TITLE
GitHub Workflows security hardening

### PR DESCRIPTION
**Pull request by sashashura**
This PR adds explicit [permissions section](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#permissions) to workflows. This is a security best practice because by default workflows run with [extended set of permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token) (except from `on: pull_request` [from external forks](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/)). By specifying any permission explicitly all others are set to none. By using the principle of least privilege the damage a compromised workflow can do (because of an [injection](https://securitylab.github.com/research/github-actions-untrusted-input/) or compromised third party tool or action) is restricted.
It is recommended to have [most strict permissions on the top level](https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions) and grant write permissions on [job level](https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs) case by case.

### CODE REVIEW

**Review by AshCorr — APPROVED** [review]
Good catch, thanks! I'll merge this down soon if nobody has any objections!

**Review by oliverlloyd — DISMISSED** [review]
Just want to check this a bit

**Comment by mxdvl** [comment]
It looks that this has prevented the `Typescript` and `TeamCity`… which means that we have no way of testing this PR against our CI pipeline.

**Comment by AshCorr** [comment]
> It looks that this has prevented the Typescript and TeamCity… which means that we have no way of testing this PR against our CI/CD pipeline.

I believe we've configured our GHA to require an approval to run if the PR is from a fork. I'm not sure how Teamcity is configured though.

**Comment by AshCorr** [comment]
Hey @sashashura,

Had a quick peek at your commit history, looks like you're using a script or something to generate this many fix PR's for so many repos? Awesome!

Just wondering, why is this permission only added to 2 of our actions? Is it not needed for the other ones?

**Comment by sashashura** [comment]
Hi @AshCorr,
Yes I have automated as much as possible. Still I review all the changes before the script creates a PR. Though it is impossible to read all contribution guides, so I apologize if my PR didn't follow a check list in case you have one.
Let me know if you have any questions about the PR.

**Comment by jfsoul** [comment]
I suspect we might want to review whether we can enable the org-wide setting for this (read-only by default), rather than harden individual workflows. Happy to suggest that to the DevX team and explore.

**Comment by sashashura** [comment]
Yes, and I encourage you doing so. Just keep in mind that any workflow that needs write permissions (like stale.yml for example), but do not have permissions explicitly specified will stop working after the org-wide switch. You might need to re-run and check if any write permission needs to be granted.

**Comment by jamesgorrie** [comment]
@jfsoul - maybe we could:
- switch it on for this repo directly
- test which actions need changing
- switch back to unsafer mode
- change them and then switch back

This would mean we're ready for a org wide switch, and might be a way to do it more incrementally.

**Comment by mxdvl** [comment]
As suggested by @akash1810, we managed to [run this in TeamCity](https://teamcity.gutools.co.uk/buildConfiguration/dotcom_master/817060) with the following script:

```sh
```

**Pull request merged**
